### PR TITLE
Fix for SR-4683

### PR DESCRIPTION
**Pull request by greglutz**
Consists of adding another case (subscripts) to the code in ASTMangler.cpp, added in 11f66f8ce39b95c62191a6291780b2316b5863bd, that fixed SR-1795.

Resolves [SR-4683](https://bugs.swift.org/browse/SR-4683).

@swift-ci Please smoke test and merge

### CODE REVIEW

**Comment by CodaFi** [comment]
@greglutz Invoking CI requires privileges.  That said, thanks for the patch, it looks great!

@swift-ci please smoke test

File: lib/AST/ASTMangler.cpp [hunk 1]
```diff
@@ -1691,14 +1691,18 @@ void ASTMangler::appendDeclType(const ValueDecl *decl, bool isFunctionMangling)
...
+    const ParameterList *Params = nullptr;
     if (const auto *FDecl = dyn_cast<AbstractFunctionDecl>(decl)) {
       unsigned PListIdx = isMethodDecl(decl) ? 1 : 0;
       if (PListIdx < FDecl->getNumParameterLists()) {
-        const ParameterList *Params = FDecl->getParameterList(PListIdx);
-        forceSingleParam = (Params->size() == 1);
+        Params = FDecl->getParameterList(PListIdx);
       }
+    } else if (const auto *SDecl = dyn_cast<SubscriptDecl>(decl)) {
+        Params = SDecl->getIndices();
```
Comment (xwu) [inline lib/AST/ASTMangler.cpp]: Drive-by nit: de-indent by two spaces?

**Comment by greglutz** [comment]
Sorry, I had it indented right initially. I think it went through
      too many busy-body editors.
    --Greg
    
    
    On 4/26/17 5:05 PM, Xiaodi Wu wrote:
    
    
      @xwu commented on this pull request.
      
      In lib/AST/ASTMangler.cpp:
      >        }
+    } else if (const auto *SDecl = dyn_cast<SubscriptDecl>(decl)) {
+        Params = SDecl->getIndices();

      Drive-by nit: de-indent by two spaces?
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/apple/swift","title":"apple/swift","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/apple/swift"}},"updates":{"snippets":[{"icon":"PERSON","message":"@xwu commented on #9046"}],"action":{"name":"View Pull Request","url":"https://github.com/apple/swift/pull/9046#pullrequestreview-34992393"}}}

**Comment by CodaFi** [comment]
@greglutz Your patch is crashing in the demangler.  Take a look at `test/SILGen/property_abstraction.swift` and
`test/SILGen/constrained_extensions.swift`.

**Comment by eeckstein** [comment]
@greglutz Your change looks good. The crashes in the demangler are caused by a bug in the mangler, which is not related to your change. I'll push a fix shortly.

**Review by eeckstein — APPROVED** [review]

**Comment by eeckstein** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-CI please smoke test

**Comment by eeckstein** [comment]
There is one test file which needs to be updated. Here is the diff:
```
diff --git a/test/SILGen/constrained_extensions.swift b/test/SILGen/constrained_extensions.swift
index d1856d00af..800b220098 100644
--- a/test/SILGen/constrained_extensions.swift
+++ b/test/SILGen/constrained_extensions.swift
@@ -48,7 +48,7 @@ extension Array where Element == Int {
     return e!
   }
 
-  // CHECK-LABEL: sil @_T0Sa22constrained_extensionsSiRszlE9subscriptSiycfg : $@convention(method) (@guaranteed Array<Int>) -> Int
+  // CHECK-LABEL: sil @_T0Sa22constrained_extensionsSiRszlE9subscriptSiyt_tcfg : $@convention(method) (@guaranteed Array<Int>) -> Int
   public subscript(i: ()) -> Element {
     return self[0]
   }
@@ -119,7 +119,7 @@ extension Dictionary where Key == Int {
     return Dictionary(x: ()).instanceMethod()
   }
 
-  // CHECK-LABEL: sil @_T0s10DictionaryV22constrained_extensionsSiRszr0_lE9subscriptq_ycfg : $@convention(method) <Key, Value where Key == Int> (@guaranteed Dictionary<Int, Value>) -> @out Value
+  // CHECK-LABEL: sil @_T0s10DictionaryV22constrained_extensionsSiRszr0_lE9subscriptq_yt_tcfg : $@convention(method) <Key, Value where Key == Int> (@guaranteed Dictionary<Int, Value>) -> @out Value
   public subscript(i: ()) -> Value {
     return self[0]!
   }
@@ -153,10 +153,10 @@ extension GenericClass where Y == () {
     set {}
   }
 
-  // CHECK-LABEL: sil @_T022constrained_extensions12GenericClassCAAytRs_r0_lE9subscriptxycfg : $@convention(method) <X, Y where Y == ()> (@guaranteed GenericClass<X, ()>) -> @out X
-  // CHECK-LABEL: sil @_T022constrained_extensions12GenericClassCAAytRs_r0_lE9subscriptxycfs : $@convention(method) <X, Y where Y == ()> (@in X, @guaranteed GenericClass<X, ()>) -> ()
-  // CHECK-LABEL: sil [transparent] [serialized] @_T022constrained_extensions12GenericClassCAAytRs_r0_lE9subscriptxycfmytfU_ : $@convention(method) <X, Y where Y == ()> (Builtin.RawPointer, @inout Builtin.UnsafeValueBuffer, @inout GenericClass<X, ()>, @thick GenericClass<X, ()>.Type) -> ()
-  // CHECK-LABEL: sil [transparent] [serialized] @_T022constrained_extensions12GenericClassCAAytRs_r0_lE9subscriptxycfm : $@convention(method) <X, Y where Y == ()> (Builtin.RawPointer, @inout Builtin.UnsafeValueBuffer, @guaranteed GenericClass<X, ()>) -> (Builtin.RawPointer, Optional<Builtin.RawPointer>)
+  // CHECK-LABEL: sil @_T022constrained_extensions12GenericClassCAAytRs_r0_lE9subscriptxyt_tcfg : $@convention(method) <X, Y where Y == ()> (@guaranteed GenericClass<X, ()>) -> @out X
+  // CHECK-LABEL: sil @_T022constrained_extensions12GenericClassCAAytRs_r0_lE9subscriptxyt_tcfs : $@convention(method) <X, Y where Y == ()> (@in X, @guaranteed GenericClass<X, ()>) -> ()
+  // CHECK-LABEL: sil [transparent] [serialized] @_T022constrained_extensions12GenericClassCAAytRs_r0_lE9subscriptxyt_tcfmytfU_ : $@convention(method) <X, Y where Y == ()> (Builtin.RawPointer, @inout Builtin.UnsafeValueBuffer, @inout GenericClass<X, ()>, @thick GenericClass<X, ()>.Type) -> ()
+  // CHECK-LABEL: sil [transparent] [serialized] @_T022constrained_extensions12GenericClassCAAytRs_r0_lE9subscriptxyt_tcfm : $@convention(method) <X, Y where Y == ()> (Builtin.RawPointer, @inout Builtin.UnsafeValueBuffer, @guaranteed GenericClass<X, ()>) -> (Builtin.RawPointer, Optional<Builtin.RawPointer>)
   public subscript(_: Y) -> X {
     get { while true {} }
     set {}
```

**Comment by greglutz** [comment]
I've managed to construct that file. Should I do the submitting? If
    so, as a new pull request, or a modification to my original one?

**Comment by CodaFi** [comment]
Just commit onto this branch, or squash and force-push.  Either way, you don't have to open a new pull request for this.

**Comment by greglutz** [comment]
Okay, I guess I've done that (committed onto my SR-4683 branch).
    Sorry for my lack of familiarity with git...

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by eeckstein** [comment]
Thanks for the fix!
